src/components/scripts/oauth-connector.ts

```diff
@@ -0,0 +1,241 @@
+  }
+
+  async #connectionReceived(connection: TCPSocket): Promise<void> {
+    return connection.opened.then(async (socket) => {
+      const reader = socket.readable.getReader();
+      const value = await this.#readStream(reader);
+      if (value == undefined) {
+        return;
+      }
+
+      const { success, html } = this.#generateResponse(value);
+      await this.#writeStream(socket, html);
+
+      connection.close();
+
+      if (success) {
+        this.#accessCode = await this.#getAccessCode();
+        if (this.#promiseResolver != undefined) {
+          this.#promiseResolver(this.#accessCode);
+        }
+      }
+    });
+  }
+
+  #generateResponse(value: AllowSharedBufferSource): {
+    success: boolean;
+    html: string;
+  } {
+    const text = new TextDecoder().decode(value);
+
+    if (this.#processRequest(text)) {
+      return {
+        success: true,
+        html: 'HTTP/1.1 200 OK\r\nContent-Type: text/html\r\n\r\n<h1>You can close this page</h1>',
+      };
+    } else {
+      return { success: false, html: 'HTTP/1.1 400 KO' };
+    }
+  }
+
+  #processRequest(text: string): boolean {
+    const lines = text.split('\r\n');
+    const params = lines[0].split(' ')[1].substring(2);
+    const urlParams = new URLSearchParams(params);
+    const state = urlParams.get('state');
+
+    if (state != null && state == 'pass-through value') {
+      const code = urlParams.get('code');
+      if (code == null) {
+        return false;
+      }
+
+      this.#authenticationCode = code;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  async #getAccessCode(): Promise<string | undefined> {
+    // See https://developers.google.com/identity/protocols/oauth2/native-appexchange-authorization-code
+
+    const url_params =
+      `?code=${this.#authenticationCode}&` +
+      `client_id=${this.#clientId}&` +
+      `client_secret=${this.#clientSecret}&` +
+      `redirect_uri=${this.#redirectUri}&` +
+      `grant_type=authorization_code`;
+
+    const url = this.#tokenEndpoint + url_params;
+
+    return fetch(url, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+    })
+      .then((response) => {
+        if (response.status >= 200 && response.status < 300) {
+          return Promise.resolve(response.json());
+        } else {
+          return Promise.reject(
+            new Error(response.status.toString + ' - ' + response.statusText),
+          );
+        }
+      })
+      .then((json) => {
+        this.#accessCode = json.access_token;
+        this.#refreshCode = json.refresh_token;
+
+        return this.#accessCode;
+      });
+  }
+
+  async #writeStream(socket: TCPSocketOpenInfo, text: string): Promise<void> {
+    const writer = socket.writable.getWriter();
+    const encoder = new TextEncoder();
+
+    const msg = encoder.encode(text);
+
+    writer.write(msg);
+    writer.releaseLock();
+  }
+
+  async #readStream(
```

Comment:
A comment that seems to have been missed from @reillyeon is that it's unsafe to always assume this will come in as one block, with a recommendation to use the [`LineBreakTransformer`](https://wicg.github.io/serial//#:~:text=class%20LineBreakTransformer) from the Web Serial spec instead.